keps/sig-network/4770-endpointslice-controller-flexibility/README.md

```diff
@@ -0,0 +1,481 @@
+
+The `SetLabelsAnnotations` function implements `SetEndpointSliceLabelsAnnotations`. The Endpoints (EndpointSlice Mirroring Controller) strategy is to set the "kubernetes.io/service-name" and "endpointslice.kubernetes.io/managed-by" labels, clones the labels and annotations from the Endpoints and updates the EndpointSlice if there is any changes.
+
+```golang
+// DesiredEndpointSlicesFromServicePods returns the list of desired endpointslices for the given endpoints.
+// It also return which address types are handled.
+func DesiredEndpointSlicesFromEndpoints(
+	endpoints *corev1.Endpoints,
+	maxEndpointsPerSubset int32,
+) ([]*EndpointPortAddressType, sets.Set[discovery.AddressType])
+
+type LabelsAnnotationsFromEndpoints struct {
+	Endpoints *corev1.Endpoints
+}
+
+func (lafe *LabelsAnnotationsFromEndpoints) SetLabelsAnnotations(logger klog.Logger, epSlice *discovery.EndpointSlice, controllerName string) (map[string]string, map[string]string, bool) 
+```
+
+Example of usage replacing the current `Reconcile` call from the EndpointSlice Mirroring Controller ([pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go#L334](https://github.com/kubernetes/kubernetes/blob/v1.30.2/pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go#L334)):
+```golang
+desiredEndpoints, supportedAddressesTypes := endpointslicerec.DesiredEndpointSlicesFromEndpoints(endpoints, r.maxEndpointsPerSubset)
+return rec.Reconcile(
+    logger,
+    endpoints,
+    endpoints,
+    desiredEndpoints,
+    existingSlices,
+    supportedAddressesTypes,
+    nil, // EndpointSlice Mirroring doesn't use traffic distribution.
+    labelsAnnotationsFromEndpoints.SetLabelsAnnotations,
+    time.Time{}, // So endpoints.kubernetes.io/last-change-trigger-time will be removed.
+)
+```
+
+#### Metrics
+
+Most of the metrics exposed by the EndpointSlice and EndpointSlice Mirroring Controllers are similar. The EndpointSlice Controller exposes `endpointslices_changed_per_sync`, `syncs` and `services_count_by_traffic_distribution` that are not exposed by the EndpointSlice Mirroring Controller. The EndpointSlice Mirroring Controller exposes `endpoints_updated_per_sync`, `addresses_skipped_per_sync` and `endpoints_sync_duration` that are not exposed by the EndpointSlice Controller. Metrics exposed will need to be merged together, and the metrics package should be changed to support any type of resources (pods/service, endpoints...).
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+* Tests in `pkg/controller/endpointslicemirroring/reconciler_test.go` will be moved to `staging/src/k8s.io/endpointslice`
+
+##### Integration tests
+
+TDB
```

Comment:
I expect to see a good battery of these tests, are the more important for this feature

Reply:
This should be fixed now. E2E tests has been moved as integration and another one as been added to tests the feature gate when disabled.